### PR TITLE
Model token

### PR DESCRIPTION
**Pull request by Ic3Sandy**

### CODE REVIEW

File: packages/bitcore-node/src/models/transaction.ts [hunk 1]
```diff
@@ -274,33 +278,67 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
...
-        let symbol = '';
-        let totalSupply = 0;
         const decimals = await rpc.call('frc20decimals', [contractAddress]);
         if (decimals !== 0) {
-          name = await rpc.call('frc20name', [contractAddress]);
-          symbol = await rpc.call('frc20symbol', [contractAddress]);
-          totalSupply = await rpc.call('frc20totalsupply', [contractAddress]);
+          const name = await rpc.call('frc20name', [contractAddress]);
+          const symbol = await rpc.call('frc20symbol', [contractAddress]);
+          const totalSupply = await rpc.call('frc20totalsupply', [contractAddress]);
```
Comment (thebevrishot) [inline packages/bitcore-node/src/models/transaction.ts]: we should call them all before consider this as a frc20

File: packages/bitcore-node/src/models/transaction.ts [hunk 2]
```diff
@@ -274,33 +278,67 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
...
+            name,
+            symbol,
+            totalSupply,
+            balances
+          };
+          TokenStorage.collection.updateOne({ txid }, { $set: token }, { upsert: true });
+          result[0].name = name;
+          result[0].decimals = decimals;
+          result[0].symbol = symbol;
+          result[0].totalSupply = totalSupply;
```
Comment (thebevrishot) [inline packages/bitcore-node/src/models/transaction.ts:311]: If this is not smallest unit please convert to before store

File: packages/bitcore-node/src/models/transaction.ts [hunk 3]
```diff
@@ -274,33 +278,67 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
...
+      ) {
+        result[0].events = [];
+        const receipt = result[0];
+        const from = receipt.log[0].topics[1].replace('000000000000000000000000', '');
+        const to = receipt.log[0].topics[2].replace('000000000000000000000000', '');
+        const value = parseInt(receipt.log[0].data, 16);
+        const contractAddress = receipt.log[0].address;
+        const token = await TokenStorage.collection.findOne({ contractAddress });
+        if (token) {
+          const balance = new Decimal(value).div(new Decimal(1e18)).toNumber();
```
Comment (thebevrishot) [inline packages/bitcore-node/src/models/transaction.ts]: We should not operate with smallest unit so we don't have to div by 1e18

File: packages/bitcore-node/src/models/transaction.ts [hunk 4]
```diff
@@ -274,33 +278,67 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
...
+          token.balances[from] -= balance;
```
Comment (thebevrishot) [inline packages/bitcore-node/src/models/transaction.ts]: please handle mint here, `from` might do not appear before

File: packages/bitcore-node/src/models/transaction.ts [hunk 5]
```diff
@@ -317,6 +355,58 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
...
+            chain,
+            network,
+            txid,
+            version: asm[asm.length - 6],
+            fvmGasLimit: asm[asm.length - 5],
+            fvmGasPrice: asm[asm.length - 4],
+            callData: asm[asm.length - 3],
+            contract: asm[asm.length - 2],
+            op: 'OP_CALL'
+          };
```
Comment (thebevrishot) [inline packages/bitcore-node/src/models/transaction.ts:404]: Are they also work with OP_CALL/OP_CREATE with signature?

File: packages/bitcore-node/src/modules/firocoin/api/firo-routes.ts [hunk 6]
```diff
@@ -30,3 +31,20 @@ FiroRoutes.get('/api/:chain/:network/prices', async (_, res) => {
...
+      .skip(+paging > 0 ? (+paging - 1) * 20 : 0)
+      .toArray();
+    for (let token of tokens) {
+      token['holders'] = Object.keys(token.balances).length;
+    }
+    res.json(tokens);
+  } catch (err) {
+    res.status(500).send(err);
+  }
+});
```
Comment (thebevrishot) [inline packages/bitcore-node/src/modules/firocoin/api/firo-routes.ts:52]: add query by address also

File: packages/bitcore-node/src/modules/firocoin/models/evmData.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  fvmGasLimit: string;
+  fvmGasPrice: number;
+  callData: string;
+  contract: string;
+  op: string;
+}
+
+export class EvmDataModel extends BaseModel<IEvmData> {
+  constructor(storage?: StorageService) {
+    super('evmdatas', storage);
```
Comment (thebevrishot) [inline packages/bitcore-node/src/modules/firocoin/models/evmData.ts]: data could not be pluralized like this `evmdata` is enough

**Review by thebevrishot — COMMENTED** [review]
What would happen when blocks reversed?

File: packages/bitcore-node/src/modules/firocoin/models/evmData.ts (hunk 7 above)
Comment (Ic3Sandy) [reply packages/bitcore-node/src/modules/firocoin/models/evmData.ts]: ok kub

File: packages/bitcore-node/src/models/transaction.ts (hunk 2 above)
Comment (Ic3Sandy) [reply packages/bitcore-node/src/models/transaction.ts:311]: ok kub

File: packages/bitcore-node/src/models/transaction.ts (hunk 1 above)
Comment (Ic3Sandy) [reply packages/bitcore-node/src/models/transaction.ts]: ok kub

File: packages/bitcore-node/src/models/transaction.ts (hunk 4 above)
Comment (Ic3Sandy) [reply packages/bitcore-node/src/models/transaction.ts]: ok kub

File: packages/bitcore-node/src/models/transaction.ts (hunk 3 above)
Comment (Ic3Sandy) [reply packages/bitcore-node/src/models/transaction.ts]: ok kub

File: packages/bitcore-node/src/modules/firocoin/api/firo-routes.ts (hunk 6 above)
Comment (Ic3Sandy) [reply packages/bitcore-node/src/modules/firocoin/api/firo-routes.ts:52]: ok kub

**Review by thebevrishot — APPROVED** [review]

**Pull request merged**